pkg/ingester/client/wrapped.go

```diff
@@ -0,0 +1,432 @@
+			}
+			if postIndex > l {
+				return io.ErrUnexpectedEOF
+			}
+			m.UserId = string(dAtA[iNdEx:postIndex])
+			iNdEx = postIndex
+		case 3:
+			if wireType != 2 {
+				return fmt.Errorf("proto: wrong wireType = %d for field Labels", wireType)
+			}
+			var msglen int
+			for shift := uint(0); ; shift += 7 {
+				if shift >= 64 {
+					return ErrIntOverflowIngester
+				}
+				if iNdEx >= l {
+					return io.ErrUnexpectedEOF
+				}
+				b := dAtA[iNdEx]
+				iNdEx++
+				msglen |= int(b&0x7F) << shift
+				if b < 0x80 {
+					break
+				}
+			}
+			if msglen < 0 {
+				return ErrInvalidLengthIngester
+			}
+			postIndex := iNdEx + msglen
+			if postIndex < 0 {
+				return ErrInvalidLengthIngester
+			}
+			if postIndex > l {
+				return io.ErrUnexpectedEOF
+			}
+			m.Labels = append(m.Labels, mimirpb.LabelAdapter{})
+			if err := m.Labels[len(m.Labels)-1].Unmarshal(dAtA[iNdEx:postIndex]); err != nil {
+				return err
+			}
+			iNdEx = postIndex
+		case 4:
+			if wireType != 2 {
+				return fmt.Errorf("proto: wrong wireType = %d for field Chunks", wireType)
+			}
+			var msglen int
+			for shift := uint(0); ; shift += 7 {
+				if shift >= 64 {
+					return ErrIntOverflowIngester
+				}
+				if iNdEx >= l {
+					return io.ErrUnexpectedEOF
+				}
+				b := dAtA[iNdEx]
+				iNdEx++
+				msglen |= int(b&0x7F) << shift
+				if b < 0x80 {
+					break
+				}
+			}
+			if msglen < 0 {
+				return ErrInvalidLengthIngester
+			}
+			postIndex := iNdEx + msglen
+			if postIndex < 0 {
+				return ErrInvalidLengthIngester
+			}
+			if postIndex > l {
+				return io.ErrUnexpectedEOF
+			}
+			// << NON AUTO-GENERATED CODE >>
+			//
+			//  m.Chunks = append(m.Chunks, Chunk{})
+			//  if err := m.Chunks[len(m.Chunks)-1].Unmarshal(dAtA[iNdEx:postIndex]); err != nil {
+			//      return err
+			//  }
+			//
+			if reusedChunks < poolChunksLength {
+				if err := m.Chunks[reusedChunks].Unmarshal(dAtA[iNdEx:postIndex]); err != nil {
+					return err
+				}
+				reusedChunks++
+			} else {
+				m.Chunks = append(m.Chunks, Chunk{})
+				if err := m.Chunks[len(m.Chunks)-1].Unmarshal(dAtA[iNdEx:postIndex]); err != nil {
+					return err
+				}
+			}
+			// << NON AUTO-GENERATED CODE >>
+			iNdEx = postIndex
+		default:
+			iNdEx = preIndex
+			skippy, err := skipIngester(dAtA[iNdEx:])
+			if err != nil {
+				return err
+			}
+			if skippy < 0 {
+				return ErrInvalidLengthIngester
+			}
+			if (iNdEx + skippy) < 0 {
+				return ErrInvalidLengthIngester
+			}
+			if (iNdEx + skippy) > l {
+				return io.ErrUnexpectedEOF
+			}
+			iNdEx += skippy
+		}
+	}
+
+	if iNdEx > l {
+		return io.ErrUnexpectedEOF
+	}
+	return nil
+}
+
+// ReuseQueryStreamResponse puts chunk and label slices contained in qr back into a sync.Pool for reuse.
+func ReuseQueryStreamResponse(qr *QueryStreamResponse) {
+	for i := 0; i < len(qr.Chunkseries); i++ {
+		if len(qr.Chunkseries[i].Chunks) > maxInPoolChunksSliceSize || len(qr.Chunkseries[i].Labels) > maxInPoolLabelsSliceSize {
```

Comment:
> maxInPoolChunksSliceSize

This value is 32K. We're saying at most we should reuse chunks slices up to 32K elements each. Chunks are per-series, so this mean we expect series from ingesters with more than 32K chunks. A chunk is full when it contains 120 samples, so 32K is about `32K * 120 samples = 3.8M samples`. Assuming a scraping interval of 10s, that's `3.8M * 10s = 439 days` of samples, but we only query the last 12h from ingesters.

Could you help me understand what we want to achieve here?

Reply:
See: https://github.com/grafana/mimir/pull/3295#discussion_r1013893742